src/H5FDmpio.c

```diff
@@ -1520,6 +1524,834 @@ H5FD__mpio_write(H5FD_t *_file, H5FD_mem_t type, hid_t H5_ATTR_UNUSED dxpl_id, h
     FUNC_LEAVE_NOAPI(ret_value)
 } /* end H5FD__mpio_write() */
 
+/*-------------------------------------------------------------------------
+ * Function:    H5FD__mpio_read_vector()
+ *
+ * Purpose:     The behaviour of this function dependes on the value of
+ *              the io_xfer_mode obtained from the context.
+ *
+ *              If it is H5FD_MPIO_COLLECTIVE, this is a collective
+ *              operation, which allows us to use MPI_File_set_view, and
+ *              then perform the entire vector read in a single MPI call.
+ *
+ *              Do this (if count is positive), by constructing memory
+ *              and file derived types from the supplied vector, using
+ *              file type to set the file view, and then reading the
+ *              the memory type from file.  Note that this read is
+ *              either independent or collective depending on the
+ *              value of mpio_coll_opt -- again obtained from the context.
+ *
+ *              If count is zero, participate in the collective read
+ *              (if so configured) with an empty read.
+ *
+ *              Finally, set the file view back to its default state.
+ *
+ *              In contrast, if io_xfer_mode is H5FD_MPIO_INDEPENDENT,
+ *              this call is independent, and thus we cannot use
+ *              MPI_File_set_view().
+ *
+ *              In this case, simply walk the vector, and issue an
+ *              independent read for each entry.
+ *
```

Comment:
Is the above description of the io_xfer_mode and the mpio_coll_opt DXPL entries correct?

Reply:
Based on the documentation for the related H5P routines, this sounds accurate to me. Particularly due to this note from H5Pset_dxpl_mpio_collective_opt:

> This function should be used only when H5FD_MPIO_COLLECTIVE has been set through H5Pset_dxpl_mpio. In such situations, normal behavior would be to use low-level collective I/O functions, but the library will use low-level MPI independent I/O functions when H5FD_MPIO_INDIVIDUAL_IO is set.

So it seems like H5Pset_dxpl_mpio for the io_xfer_mode is for specifying collectivity at the application level, while H5Pset_dxpl_mpio_collective_opt is available to ask HDF5 to actually use independent I/O regardless of the application level collectivity setting.